### PR TITLE
Rust nightly removed the lifetime from Pattern

### PR DESCRIPTION
**Pull request by mischnic**
Closes https://github.com/rust-lang/regex/issues/1216

### CODE REVIEW

File: src/pattern.rs [hunk 1]
```diff
@@ -10,10 +10,10 @@ pub struct RegexSearcher<'r, 't> {
     next_match: Option<(usize, usize)>,
 }
 
-impl<'r, 't> Pattern<'t> for &'r Regex {
-    type Searcher = RegexSearcher<'r, 't>;
+impl<'r> Pattern for &'r Regex {
+    type Searcher<'t> = RegexSearcher<'r, 't>;
 
-    fn into_searcher(self, haystack: &'t str) -> RegexSearcher<'r, 't> {
+    fn into_searcher(self, haystack: &str) -> RegexSearcher<'r, '_> {
```
Comment (BurntSushi) [inline src/pattern.rs:16]: Should this be?

```
fn into_searcher(self, haystack: &'t str) -> RegexSearcher<'r, 't> {
```

Comment (mischnic) [reply src/pattern.rs:16]: This lifetime setup is also used by the various `Pattern` impls in the standard library:
https://github.com/rust-lang/rust/commit/772315de7c1f68bb95de5a7eb85cb2f167834c53

Your version doesn't compile directly, the first suggestion below is exactly what I have implemented now (but with `'t` being elided) and the second suggestion is what was used before my PR and now fails to compile.
```
error[E0261]: use of undeclared lifetime name `'t`
  --> src/pattern.rs:16:68
   |
16 |     fn into_searcher(self, haystack: &'t str) -> RegexSearcher<'r, 't> {
   |                                                                    ^^ undeclared lifetime
   |
help: consider introducing lifetime `'t` here
   |
16 |     fn into_searcher<'t>(self, haystack: &'t str) -> RegexSearcher<'r, 't> {
   |                     ++++
help: consider introducing lifetime `'t` here
   |
13 | impl<'t, 'r> Pattern for &'r Regex {
   |      +++
```

Comment (BurntSushi) [reply src/pattern.rs:16]: I guess the problem I have here is how does the `RegexSearcher` connect its lifetime to the haystack? I must be missing something.

Comment (mischnic) [reply src/pattern.rs:16]: What I wrote here is equivalent to this, which is how the `&str` lifetime flows into the `RegexSearcher` lifetime (I can commit that instead if you want)
```
fn into_searcher<'h>(self, haystack: &'h str) -> RegexSearcher<'r, 'h>
```

Comment (BurntSushi) [reply src/pattern.rs:16]: Ooooo, yes, I get it now. Yes, please do write that. I'm usually fine with lifetime elision, but I think it actually helps a lot here.

Comment (BurntSushi) [reply src/pattern.rs:16]: Right right. This all makes sense now. Got it. I was confused myself.

Comment (mischnic) [reply src/pattern.rs]: Perfect 😄 
I've added back that explicit lifetime

File: src/pattern.rs [hunk 2]
```diff
@@ -10,10 +10,10 @@ pub struct RegexSearcher<'r, 't> {
     next_match: Option<(usize, usize)>,
 }
 
-impl<'r, 't> Pattern<'t> for &'r Regex {
-    type Searcher = RegexSearcher<'r, 't>;
+impl<'r> Pattern for &'r Regex {
+    type Searcher<'t> = RegexSearcher<'r, 't>;
 
-    fn into_searcher(self, haystack: &'t str) -> RegexSearcher<'r, 't> {
+    fn into_searcher<'h>(self, haystack: &'h str) -> RegexSearcher<'r, 'h> {
```
Comment (BurntSushi) [inline src/pattern.rs:16]: Sorry, can this be `'t` instead of `'h` to be consistent? I don't think there is a shadowing concern here.

**Review by BurntSushi — APPROVED** [review]
Perfect, thanks!

**Comment by BurntSushi** [comment]
This fix is on crates.io in `regex 1.10.6`.

**Comment by mischnic** [comment]
Great, thank you for the fast release!

**Pull request merged**